symbolic-base/src/ZkFold/Base/Algebra/Basic/Field.hs

```diff
@@ -226,48 +226,51 @@ instance Field f => AdditiveGroup (Ext2 f e) where
     negate (Ext2 a b) = Ext2 (negate a) (negate b)
     Ext2 a b - Ext2 c d = Ext2 (a - c) (b - d)
 
-instance {-# OVERLAPPING #-} (Field f, Eq f, IrreduciblePoly f e) => Scale (Ext2 f e) (Ext2 f e)
+instance {-# OVERLAPPING #-} (Field f, Eq f, IrreduciblePoly poly f e) => Scale (Ext2 f e) (Ext2 f e)
 
-instance (Field f, Eq f, IrreduciblePoly f e) => MultiplicativeSemigroup (Ext2 f e) where
-    Ext2 a b * Ext2 c d = fromConstant (toPoly [a, b] * toPoly [c, d])
+instance (Field f, Eq f, IrreduciblePoly poly f e) => MultiplicativeSemigroup (Ext2 f e) where
+    Ext2 a b * Ext2 c d = fromConstant @poly (toPoly @f @poly [a, b] * toPoly @f @poly [c, d])
 
 instance MultiplicativeMonoid (Ext2 f e) => Exponent (Ext2 f e) Natural where
     (^) = natPow
 
-instance (Field f, Eq f, IrreduciblePoly f e) => MultiplicativeMonoid (Ext2 f e) where
+instance (Field f, Eq f, IrreduciblePoly poly f e) => MultiplicativeMonoid (Ext2 f e) where
     one = Ext2 one zero
 
 instance Field (Ext2 f e) => Exponent (Ext2 f e) Integer where
     (^) = intPowF
 
-instance (Field f, Eq f, IrreduciblePoly f e) => Field (Ext2 f e) where
+instance (Field f, Eq f, IrreduciblePoly poly f e) => Field (Ext2 f e) where
     finv (Ext2 a b) =
-        let (g, s) = eea (toPoly [a, b]) (irreduciblePoly @f @e)
+        let (g, s) = eea (toPoly [a, b]) (irreduciblePoly @poly @f @e)
         in case fromPoly $ scaleP (one // lt g) 0 s of
             []  -> Ext2 zero zero
             [x] -> Ext2 x zero
             v   -> Ext2 (v V.! 0) (v V.! 1)
 
     rootOfUnity n = (\r -> Ext2 r zero) <$> rootOfUnity n
 
-instance (FromConstant f f', Field f') => FromConstant f (Ext2 f' e) where
-    fromConstant e = Ext2 (fromConstant e) zero
+instance {-# OVERLAPPING #-} (Field f, IrreduciblePoly poly f e) => FromConstant Natural (Ext2 f e) where
+    fromConstant = flip scale one
 
-instance {-# OVERLAPPING #-} (Field f, Eq f, IrreduciblePoly f e) => FromConstant (Poly f) (Ext2 f e) where
-    fromConstant p = case fromPoly . snd $ qr p (irreduciblePoly @f @e) of
+instance {-# OVERLAPPING #-} (Field f, IrreduciblePoly poly f e) => FromConstant Integer (Ext2 f e) where
```

Comment:
I think the `FromConstant Natural` and `FromConstant Integer` instances can be removed since we have a `FromConstant c` instance which covers its cases